runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,481 @@
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "chplrt.h"
+#include "chpl-thread-local-storage.h"
+#include "chpl-mem.h"
+#include "chpl-atomics.h"
+#include "chpl-tasks.h"
+#include "chpl-qsbr.h"
+
+// Number of checkpoints to pass before processing
+#define CHPL_QSBR_CHECKPOINTS_PER_PROCESS 8
+
+
+
+static atomic_uint_least64_t global_epoch = 0;
+
+// Each node in the deferred list has associated with it
+// the data (and numData, useful for determining if it
+// is an array and should be treated as a 'void **') and
+// the targetEpoch which is the epoch required for all threads
+// to have observed before the data can be deleted. By making
+// use of the fact that (barring integer overflow) the largest
+// targetEpoch is always pushed as the head, we have a relative
+// sorting order. If we ask for some targetEpoch X, and for some
+// node N we have X >= N.targetEpoch, then we can claim N and
+// N.next and so and so forth since N.targetEpoch >= N.next.targetEpoch. 
+struct defer_node {
+  uint64_t targetEpoch;
+  void *data;
+  int numData;
+  struct defer_node *next;
+};
+
+// Thread-local data that keeps track of our observed global state.
+struct tls_node {
+  /*
+    The number of checkpoints passed through
+  */
+  uint64_t passedCheckpoints;
+  /*
+    The current observed epoch. This is used to ensure that writers
+    do not delete an older instance that corresponds to this epoch.
+  */
+  atomic_uint_least64_t epoch;
+  /*
+    Number of tasks that this thread has running. Required so
+    that threads without tasks (and no longer are able to call
+    checkpoints) are not taken into account.
+  */
+  atomic_uint_least64_t nTasks;
+  /*
+    List of objects that are queued for deferred deletion, sorted by target_epoch. 
+    We defer deletion if we find that while scanning the list of all thread-specific
+    data, they are not up-to-date on that epoch. By sorting by target_epoch, we can
+    make shortcuts by removing a large bulk of data at once. The list is protected
+    by a lightweight spinlock, but is acquired using pseudo-TestAndSet operations
+    due to low contention.
+  */
+  atomic_uint_least64_t spinlock;
+  struct defer_node *deferList;
+
+  /*
+    Recycled defer nodes that can be reused.
+  */
+  struct defer_node *recycleList;
+  struct tls_node *next;
+};
+
+// List of thread-local data.
+static struct tls_node *chpl_qsbr_tls_list;
+static CHPL_TLS_DECL(struct tls_node *, chpl_qsbr_tls);
+
+static inline struct tls_node *get_tls_list(void);
+static inline struct tls_node *get_tls_list(void) {
```

Comment:
Might want to mention that handling it this way (vs caching it in TLS, say) allows for the case that threads are added partway through a computation.

Reply:
Honestly I was thinking of finding a way to initialize all threads TLS at initialization time, but I wasn't sure how to do this. However for now I'll add some extra documentation along those lines.